shell/src/context_listener.rs

```diff
@@ -195,12 +205,54 @@ fn listen_protocol_events(
     Ok(())
 }
 
-fn perform_context_action(
+// returns hash of the merkle tree that action needs to be
+// applied on
+pub fn get_tree_hash(action: &ContextAction) -> Option<[u8;32]> {
+    match &action
+    {
+        ContextAction::Get {tree_hash, .. }
+        | ContextAction::Mem {tree_hash, .. }
+        | ContextAction::DirMem {tree_hash, ..}
+        | ContextAction::Set {tree_hash, ..}
+        | ContextAction::Copy {tree_hash, ..}
+        | ContextAction::Delete {tree_hash, ..}
+        | ContextAction::RemoveRecursively {tree_hash, ..}
+        | ContextAction::Commit {tree_hash, ..}
+        | ContextAction::Fold {tree_hash, ..} => {
+            let mut hash:[u8;32] = [0;32];
+            tree_hash.clone().reader().read_exact(& mut hash).unwrap();
+            Some(hash)
+        }
+        _ => {None}
+    }
+}
+
+pub fn get_new_tree_hash(action: &ContextAction) -> Option<[u8;32]> {
+    match &action{
+        ContextAction::Set {new_tree_hash, ..}
+        | ContextAction::Copy {new_tree_hash, ..}
+        | ContextAction::Delete {new_tree_hash, ..}
+        | ContextAction::RemoveRecursively {new_tree_hash, ..} => {
+            let mut hash:[u8;32] = [0;32];
+            new_tree_hash.clone().reader().read_exact(& mut hash).unwrap();
+            Some(hash)
+        }
+        _ => {None}
+    }
+}
+
+
+pub fn perform_context_action(
     action: &ContextAction,
     context: &mut Box<dyn ContextApi>,
 ) -> Result<(), Error> {
+
+    if let Some(pre_hash) = get_tree_hash(&action){
+        context.set_merkle_root(pre_hash)?;
+    }
+    
     match action {
-        ContextAction::Get { key, .. } => {
+        ContextAction::Get { key,tree_hash, .. } => {
```

Comment:
this `tree_hash` is not being used for anything right? it is handled above if I'm not misunderstanding